File: Day 3/AOC2023_3_P2.rb
class Information
    attr_accessor :line, :start, :end_c, :is_symbol, :value
    # Represents the infos of a number or a symbol (its value, and coordinates)
    def initialize(val, line, start, e, symb)
        @value = val
        @line = line
        @start = start
        @end_c = e
        @is_symbol = symb
    end
    def to_s()
        return "Value : #{@value}, line : #{@line}, begin : #{@start}, end : #{@end_c}, symbol? : #{@is_symbol}"
    end
end

def checkAround(list, info)
    # Checks the coordinates around (square) if there is a symbol not "."
    found = []
    list.each do |elt|
        coord_elt = (elt.start - 1 .. elt.end_c + 1)
        if not elt.is_symbol and coord_elt.member?(info.start)
            found.append(elt.value.to_i)
        end
    end
    if found.size == 2
        return found[0] * found[1]
    end
    return 0
end

def processLine(line_nb, line_content)
    # We change the "*" and "$" makes the regex manipulation harder

    # Creates Information for each relevant item in the line
    arr_info = []
    line_content.scan(/\d+|[^a-z0-9.\s]/).each do |x|
        infos = line_content.match(Regexp.escape(x))
        arr_info << Information.new(x, line_nb, infos.begin(0), infos.end(0) - 1, x.match?(/[^a-z0-9.\s]/))
        line_content.sub!(x, "."*x.size)
    end
    return arr_info
end

infos = []
File.open("inputs").each_with_index do |x, i|
    infos += (processLine(i, x))
end 

i = 0
infos.each do |elt|
    if elt.is_symbol && elt.value =~ /\*/
        i += checkAround(infos.select{|x| (elt.line - 1 .. elt.line + 1).member?(x.line)}, elt)
    end
end
puts i

File: Day 3/AOC2023_3_P2.nim
import std/nre
import std/strutils
import std/sugar
import sequtils
import std/math

type
    Information* = object of RootObj
        value : string
        line : int
        start : int
        fin : int
        isSymbol : bool

proc createInformation(value : string, line, start, fin : int, isSymbol : bool) : Information =
    ## Creates an Information object
    Information(value : value, line: line, start : start, fin : fin, isSymbol : isSymbol)

proc checkAround(info : Information, around : openArray[Information]) : int =
    var found : seq[Information]
    for elt in around:
        let coord = elt.start - 1 .. elt.fin + 1
        if not elt.isSymbol and info.start in coord:
            found.add(elt)
    if found.len == 2:
        return found[0].value.parseInt * found[1].value.parseInt
    else:
        return 0

proc processLine(line_nb : int, line_content : string) : seq[Information] =
    for elt in findIter(line_content, re"\d+|[^a-z0-9.\s]"):
        let symb = elt.match.contains(re"[^a-z0-9.\s]")
        result.add(createInformation(elt.match, line_nb, elt.matchBounds.a, elt.matchBounds.b, symb))

var infos_list: seq[Information]
var line = 0
for elt in lines("inputs"):
    infos_list.add(processLine(line, elt))
    line += 1

let res = collect:
        for elt in infos_list:
            if elt.isSymbol and elt.value == "*":
                checkAround(elt, infos_list.filter( x => x.line in elt.line - 1 .. elt.line + 1))
echo sum(res)